### PR TITLE
test(mapper): add thrift error and helper fuzz tests

### PR DESCRIPTION
**Pull request by AkashKumar7902**
**What changed?**

Added gofuzz-based round-trip tests for thrift error mappers and thrift helper time/nanosecond conversions. Refs #7611.

**Why?**

The mapper fuzz test tracking issue lists errors and helpers as remaining coverage areas. These tests exercise randomized mapper inputs, preserve nil edge cases, and constrain time/error values where the existing conversions require paired fields or safe timestamp ranges.

**How did you test it?**

- `go test ./common/types/mapper/thrift -run 'Test(ErrorsFuzz|TimeToNanoFuzz|NanoToTimeFuzz|Errors|NilMapsToNil|FromUnknownErrorMapsToItself|ToUnknownErrorMapsToItself)$' -count=1`
- `go test ./common/types/mapper/thrift -count=1`
- `go test ./common/types/mapper/... -count=1`
- `git diff --check`

**Potential risks**

Low. This is test-only and does not change mapper behavior.

**Release notes**

N/A. Test-only change for #7611.

**Documentation Changes**

N/A.

### CODE REVIEW

**Comment by AkashKumar7902** [comment]
Resolved the merge conflict by rebasing onto current `master`; the conflict was in `helpers_test.go`, and the branch now keeps both the newer upstream helper conversion tests and these fuzz tests. Local validation passed with `go test ./common/types/mapper/thrift` and `git diff --check`.

File: common/types/mapper/thrift/errors_test.go [hunk 1]
```diff
@@ -55,3 +59,40 @@ func TestToUnknownErrorMapsToItself(t *testing.T) {
...
+				customFuncs = []interface{}{RetryTaskV2ErrorFuzzer}
+			}
+			f := fuzz.NewWithSeed(seed).Funcs(customFuncs...)
+			for i := 0; i < testutils.DefaultIterations; i++ {
+				newErr := reflect.New(errType).Interface()
+				f.Fuzz(newErr)
+				assert.Equal(t, newErr, ToError(FromError(newErr.(error))), "iteration %d", i)
+			}
+		})
+	}
```
Comment (c-warren) [inline common/types/mapper/thrift/errors_test.go:97]: This looks very similar to `RunMapperFuzzTest` from our testutils helpers, which you're also using below. What is the reason that we need this custom behavior instead?

File: common/types/mapper/thrift/helpers_test.go [hunk 2]
```diff
@@ -91,3 +94,23 @@ func TestSecondsI32ToDurationNil(t *testing.T) {
 	result := secondsToDuration(nil)
 	assert.Equal(t, time.Duration(0), result)
 }
+
+func SafeNanoFuzzer(n *int64, c fuzz.Continue) {
+	*n = c.Int63n(int64(testutils.MaxSafeTimestampSeconds) * int64(testutils.NanosecondsPerSecond))
+}
```
Comment (c-warren) [inline common/types/mapper/thrift/helpers_test.go:100]: NIT: All of our functions should be Safe to run - I know we already have this in the repo, but I'd prefer to name functions so it's clear how/why they exist. If we keep it I'd recommend updating the name so that it's clear that its generating values compatible with both int64 and time.Time.

File: common/types/mapper/thrift/helpers_test.go [hunk 3]
```diff
@@ -91,3 +94,23 @@ func TestSecondsI32ToDurationNil(t *testing.T) {
...
+	testutils.RunMapperFuzzTest(t, timeToNano, nanoToTime,
+		testutils.WithCustomFuncs(LocalTimeFuzzer),
+	)
+}
+
+func TestNanoToTimeFuzz(t *testing.T) {
+	testutils.RunMapperFuzzTest(t, nanoToTime, timeToNano,
+		testutils.WithCustomFuncs(SafeNanoFuzzer),
+	)
+}
```
Comment (c-warren) [inline common/types/mapper/thrift/helpers_test.go:116]: I think fuzz testing TimeToNano might not be worthwhile - it's a single field, and the edge cases are relatively clear - time.Time max, time.Time min, negative, 0 value, nil values. We're covering 4 of those already, I'd recommend instead of fuzzing we hit the other edge cases. 

Fuzz testing is naturally more expensive (as we have to run the test 100s of times to try and find edge cases that aren't covered).

Comment (c-warren) [reply common/types/mapper/thrift/helpers_test.go:116]: Something like this:

```
timeToNano / nanoToTime
- TestTimeToNanoUnixEpoch — time.Unix(0,0) → int64(0)
- TestTimeToNanoGoZeroTime — time.Time{} returns non-nil 
- TestTimeToNanoNegative — pre-epoch time → negative nanos
- TestNanoToTimeEpoch — int64(0) → epoch
- TestNanoToTimeNegative — negative nanos → pre-epoch time
- TestTimeToNanoRoundTrip

timeValToNano / nanoToTimeVal
- TestTimeValToNanoUnixEpoch — epoch's IsZero() is false, so returns &int64(0), not nil
- TestTimeValToNanoNegative — pre-epoch time
- TestNanoToTimeValEpoch — int64(0) → epoch (not zero time)
- TestNanoToTimeValNegative — negative nanos
- TestTimeValToNanoRoundTrip
```

**Review by zawadzkidiana — COMMENTED** [review]
disregard

**Comment by dianazawadzki** [comment]
> Nice initiative picking up more of #7611's remaining coverage — round-trip fuzzing the thrift error mapper and the time/nano helpers is exactly the kind of test this codebase asks for (`ToX(FromX(item)) == item`, per the `common/types/mapper/proto/api_test.go` pattern). The `RetryTaskV2ErrorFuzzer` correctly preserves the paired-nullness invariant on `StartEventID`/`StartEventVersion` and `EndEventID`/`EndEventVersion` (if you fuzzed those independently you'd get spurious failures on inputs the real type never produces), and seeding with `time.Now().UnixNano()` while logging the seed on failure is a good touch for reproducing a flake.
> 
> I noticed @c-warren left three review comments back in June that don't look like they've been addressed yet (branch is still at the original single commit from April):
> 
> 1. Why does the error-mapper test need its own custom fuzz loop instead of reusing `testutils.RunMapperFuzzTest`, which the helpers tests already use?
> 2. `SafeNanoFuzzer` — naming nit, since apparently all fuzzers in this repo are expected to be "safe"; worth a name that says what makes this one specific (int64-and-time.Time-range-compatible).
> 3. Whether `TimeToNano`/`NanoToTime` are worth fuzzing at all vs. just adding the handful of explicit edge cases (epoch, negative/pre-epoch, zero-value) c-warren listed out — fuzzing a single scalar field runs hundreds of iterations to rediscover edges that are already enumerable by hand.
> 
> On (3) in particular, I lean toward agreeing explicit edge-case tests would be more informative here for `TimeToNano`/`NanoToTime` specifically, while the error-mapper fuzz test (which has real combinatorial nested-struct surface) is a good fit for fuzzing as originally written. Would be great to see this land — mind following up on that feedback so it can move forward?
> 
> _Generated by [Claude Code](https://claude.ai/code)_

working on Human review -- my aplogies!
Diana

**Review by dianazawadzki — CHANGES_REQUESTED** [review]
Still relevant, none of this is on master. Blocked on c-warren's three from
June 5, branch hasn't moved since April.

1. RunMapperFuzzTest is at testutils/fuzz_mapper.go:132. Say why the errors
   test needs its own loop (paired nullness on RetryTaskV2Error?) or switch to
   it.
2. Rename SafeNanoFuzzer to say what it actually constrains.
3. Agree with c-warren here, take the edge cases over the fuzz. master only
   covers positive and nil for all four helpers. Epoch, pre-epoch and round
   trip are all missing, and timeValToNano(time.Unix(0,0)) is the one worth
   pinning since IsZero() is false there so it returns &0, not nil.

The errors fuzz test is fine as is, it has real nested surface. It's the
single-field time helpers that don't need it.
